GPU/GPUTracking/Merger/GPUTPCGMMerger.cxx

```diff
@@ -502,8 +502,11 @@ void GPUTPCGMMerger::MergeBorderTracks(int iSlice1, GPUTPCGMBorderTrack B1[], in
         if (b2.NClusters() < lBest2) {
           CADEBUG2(continue, printf("!NCl1\n"));
         }
-        if (crossCE >= 2 && abs(b1.Row() - b2.Row()) > 1) {
-          CADEBUG2(continue, printf("!ROW\n"));
+        if (crossCE) {
+          int maxRowDiff = crossCE == 2 ? 1 : 3; // TODO: check cut
+          if (abs(b1.Row() - b2.Row()) > maxRowDiff) {
```

Comment:
I think one can be more selective than abs diff of rows: if the track on C side has tgL>0, then I would request ``0<=rowA - rowC<=maxRowDiff`` and
``0<=rowC - rowA<=maxRowDiff`` for tgL<0.

I did not understand how to select here which one is A or C track (sign of Z coordinate sometimes is the same for both tracks). But I see something which I don't understand: I moved this Rows check to the very end of tracks comparison (so that only tracks passing all chi2... selections are checked for rows difference) and dumped tracks params and rows, i.e.
````
--- a/GPU/GPUTracking/Merger/GPUTPCGMMerger.cxx
+++ b/GPU/GPUTracking/Merger/GPUTPCGMMerger.cxx
@@ -502,12 +502,6 @@ void GPUTPCGMMerger::MergeBorderTracks(int iSlice1, GPUTPCGMBorderTrack B1[], in
         if (b2.NClusters() < lBest2) {
           CADEBUG2(continue, printf("!NCl1\n"));
         }
-        if (crossCE) {
-          int maxRowDiff = crossCE == 2 ? 1 : 3; // TODO: check cut
-          if (abs(b1.Row() - b2.Row()) > maxRowDiff) {
-            CADEBUG2(continue, printf("!ROW\n"));
-          }
-        }
         if (!b1.CheckChi2Y(b2, factor2ys)) {
           CADEBUG2(continue, printf("!Y\n"));
         }
@@ -531,7 +525,20 @@ void GPUTPCGMMerger::MergeBorderTracks(int iSlice1, GPUTPCGMBorderTrack B1[], in
             CADEBUG2(continue, printf("!NCl3\n"));
           }
         }
-        CADEBUG(printf("OK: dZ %8.3f D1 %8.3f D2 %8.3f\n", fabsf(b1.Par()[1] - b2.Par()[1]), 3.5 * sqrt(b1.Cov()[1]), 3.5 * sqrt(b2.Cov()[1])));
+
+        if (crossCE) {
+          int maxRowDiff = crossCE == 2 ? 1 : 3; // TODO: check cut
+         for (int ip=0;ip<5;ip++) {
+           printf("[%d] %+.2f/%+.2f ",ip,b1.Par()[ip], b2.Par()[ip]);
+         }
+         printf(" Rows:%d/%d MaxDR:%d -> %d\n",b1.Row(),b2.Row(), maxRowDiff, abs(b1.Row() - b2.Row()) <= maxRowDiff);
+          if (abs(b1.Row() - b2.Row()) > maxRowDiff) {
+            CADEBUG2(continue, printf("!ROW\n"));
+          }
+        }
+
+
+       CADEBUG(printf("OK: dZ %8.3f D1 %8.3f D2 %8.3f\n", fabsf(b1.Par()[1] - b2.Par()[1]), 3.5 * sqrt(b1.Cov()[1]), 3.5 * sqrt(b2.Cov()[1])));
       } // DEBUG CODE, match by MC label
       lBest2 = b2.NClusters();
       iBest2 = b2.TrackID();
````
What I see is that most of track pairs passing selection have Rows=0, event thos the Y coordinates are suggest much larger radius (if I interpret it correctly)
````
[31935:tpc-tracker]: [0] +114.09/+114.11 [1] +2.60/-4.88 [2] +0.79/+0.78 [3] -0.19/-0.19 [4] -8.59/-7.98  Rows:0/0 MaxDR:3 -> 1
[31935:tpc-tracker]: [0] +205.72/+204.30 [1] +15.95/-8.69 [2] -0.33/-0.30 [3] -0.25/-0.25 [4] +9.86/+9.39  Rows:0/0 MaxDR:3 -> 1
[31935:tpc-tracker]: [0] +132.19/+130.56 [1] -31.64/+16.81 [2] -0.86/-0.86 [3] +0.84/+0.82 [4] -4.80/-5.44  Rows:0/0 MaxDR:3 -> 1
[31935:tpc-tracker]: [0] +107.70/+107.69 [1] -9.01/+4.34 [2] -0.69/-0.64 [3] +0.90/+0.90 [4] -8.77/-7.78  Rows:0/0 MaxDR:3 -> 1
````
The dump is here: https://cernbox.cern.ch/index.php/s/OKvejOpMFUyefhE
What did I miss?

Reply:
Indeed, one could be more selective. I even though that one should check the difference in the absolute time of the cluster closest to the CE, and then accommodate for the difference based on track sinPhi and tgl. But for now, I didn't want the cut too strict.

Row = 0 indeed means tracks which start at row = 0, therefore there is an abundance of such cases. I filtered them out at first (tracks with row < 5 and row > NROWS - 5) but then I loost CE-crossing looping tracks at upper / lower boundary of the TPC. I guess I should add this cut again for tracks > 200 MeV or so.

row and y is not strictly related. I compare the tracks twice:
- at paramters of the row where they (supposedly) reach the CE.
- at Row 63 (end of IROC).
The problem with the first comparison is that it will often fail, when the clusters closest to CE are at different rows, so I added the second comparison at Row 63. (I do not want to propagate during the comparison).
These are the cases where you see row 0 and larger y than what seems possible.

Reply:
@davidrohr thanks for explanations, as far as I understand it, CE crossing for at row~0 or row~151 is possible only for loopes reaching at CE their min or max R respectively. In this case they should typically have large snp params (i.e. track almost parallel to the padrow) of opposite sign. I still did not understand why so many  tracks pass matching with both ate row=0: if this is the row at which the track is supposed to cross the CE, except loopers there should be no such tracks.
Also, I did not understand the meaning of calling MergeBorderTracks with crossCE=-1 for neighbouring sectors of the same side. If this is just to trigger the comparison of rows, why this is not done always?
Anyway, perhaps it is better to discuss this near the whiteboard...

Reply:
> Also, I did not understand the meaning of calling MergeBorderTracks with crossCE=-1 for neighbouring sectors of the same side. If this is just to trigger the comparison of rows, why this is not done always?

That is actually wrong. I just abused the crossCE to test some special treatment for the additional looper iterations. I just renamed it.

For your printout, please use
```
+        if (mergeMode > 0) {
+          int maxRowDiff = mergeMode == 2 ? 1 : 3; // TODO: check cut
+          for (int ip = 0; ip < 5; ip++) {
+            printf("[%d] %+.2f/%+.2f ", ip, b1.Par()[ip], b2.Par()[ip] + (ip == 1 ? b2.ZOffset() - b1.ZOffset() : 0.f));
+          }
+          printf(" Rows:%d/%d MaxDR:%d -> %d\n", (int)b1.Row(), (int)b2.Row(), maxRowDiff, abs(b1.Row() - b2.Row()) <= maxRowDiff);
+          if (abs(b1.Row() - b2.Row()) > maxRowDiff) {
+            CADEBUG2(continue, printf("!ROW\n"));
+          }
+        }
```

So you printed many row0 candidates, which were not crossing CE.
Still, there are plenty of matches which probably should not be merged.
We definitely need also some stricter cut on SinPhi and tgl, despite large cov. I think it could be the very short track segments reaching in an adjacent sector which hurt us here. Perhaps I could use the parameters from the long segment that was used for the extrapolation instead.

Reply:
OK, let's discuss tomorrow. One more (general) point: could you substitute all cuts on track->QPt() or track->Par()[4] by cuts on q/pt * B/5kG to make it valid also for the reduced field?